# The Purpose of this Notebook is to Scrape Web Data from the Toronto Wikipedia Page and Append Lattitude and Longitude Data

In [1]:
from urllib.request import urlopen
from bs4 import BeautifulSoup
import ssl
import pandas as pd

In [2]:
ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
html = urlopen(url, context=ctx).read() #opens the web page and reads it
soup = BeautifulSoup(html, "html.parser") 

In [3]:
table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

# print(table_contents)
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

In [4]:
df.head(10)

PostalCode           Borough                      Neighborhood
0        M3A        North York                         Parkwoods
1        M4A        North York                  Victoria Village
2        M5A  Downtown Toronto         Regent Park, Harbourfront
3        M6A        North York  Lawrence Manor, Lawrence Heights
4        M7A      Queen's Park     Ontario Provincial Government
5        M9A         Etobicoke                  Islington Avenue
6        M1B       Scarborough                    Malvern, Rouge
7        M3B        North York                   Don Mills North
8        M4B         East York   Parkview Hill, Woodbine Gardens
9        M5B  Downtown Toronto          Garden District, Ryerson

In [5]:
df.shape

(103, 3)

The dataframe created from the wikipedia table consists of 103 rows and 3 columns, plus the index column

## In this section, the latitude and longitude data will be appended to the dataframe

In [6]:
geoloc = pd.read_csv("GeoCoords.csv")
print(geoloc.head())

  Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476


In [7]:
df.set_index('PostalCode', inplace = True)
geoloc.set_index('Postal Code', inplace = True)

In [8]:
joindf = pd.DataFrame(pd.merge(df, geoloc, left_index = True, right_index = True))

In [9]:
joindf.head()

Borough                      Neighborhood   Latitude  Longitude
M3A        North York                         Parkwoods  43.753259 -79.329656
M4A        North York                  Victoria Village  43.725882 -79.315572
M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260 -79.360636
M6A        North York  Lawrence Manor, Lawrence Heights  43.718518 -79.464763
M7A      Queen's Park     Ontario Provincial Government  43.662301 -79.389494

In [10]:
joindf.reset_index(inplace = True)
joindf.head()

index           Borough                      Neighborhood   Latitude  \
0   M3A        North York                         Parkwoods  43.753259   
1   M4A        North York                  Victoria Village  43.725882   
2   M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3   M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4   M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

In [11]:
import folium
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

In [12]:
address = 'Toronto, Ontario, Canada'

geolocator = Nominatim(user_agent="TO_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [13]:
# create map of Toronto using latitude and longitude values
map_TO = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(joindf['Latitude'], joindf['Longitude'], joindf['Borough'], joindf['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_TO)  
    
map_TO

In [14]:
#@ hidden_cell
CLIENT_ID = 'LFF4YRZCAKT4UWBR5DRZQRS3OR4MEJQFX1DVG0ZGKS1SQULI' # your Foursquare ID
CLIENT_SECRET = 'OXVIJ110MU3QNYDR2IGDZ2G1IH1IDB44IYEAP5UTWAEZTUP2' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value
import json
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

### Now lets get the venue information in Toronto

In [15]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [16]:
Toronto_venues = getNearbyVenues(names=joindf['Neighborhood'],
                                   latitudes=joindf['Latitude'],
                                   longitudes=joindf['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills North
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview East
The Danforth

In [17]:
print(Toronto_venues.shape)
Toronto_venues.head()

(1984, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3         Parkwoods              43.753259              -79.329656   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0                     KFC       43.754387       -79.333021   
1         Brookbanks Park       43.751976       -79.332140   
2     Towns On The Ravine       43.754754       -79.332552   
3           Variety Store       43.751974       -79.333114   
4  Victoria Village Arena       43.723481       -79.315635   

         Venue Category  
0  Fast Food Restaurant  
1                  Park  
2                 Hotel  
3     Food & Drink Shop  
4          Hockey Arena

In [18]:
Toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                             
Agincourt                                                            4   
Alderwood, Long Branch                                               7   
Bathurst Manor, Wilson Heights, Downsview North                     20   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   21   
...                                                                ...   
Willowdale West                                                      6   
Willowdale, Newtonbrook                                              2   
Woburn                                                               4   
Woodbine Heights                                                     5   
York Mills West                                                      2   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Agincourt                                                             4   
Alderwood, Long Branch                                                7   
Bathurst Manor, Wilson Heights, Downsview North                      20   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    21   
...                                                                 ...   
Willowdale West                                                       6   
Willowdale, Newtonbrook                                               2   
Woburn                                                                4   
Woodbine Heights                                                      5   
York Mills West                                                       2   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Agincourt                                            4               4   
Alderwood, Long Branch                               7               7   
Bathurst Manor, Wilson Heights, Downsview North     20              20   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   21              21   
...                                                ...             ...   
Willowdale West                                      6               6   
Willowdale, Newtonbrook                              2               2   
Woburn                                               4               4   
Woodbine Heights                                     5               5   
York Mills West                                      2               2   

                                                 Venue Longitude  \
Neighborhood                                                       
Agincourt                                                      4   
Alderwood, Long Branch                                         7   
Bathurst Manor, Wilson Heights, Downsview North               20   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             21   
...                                                          ...   
Willowdale West                                                6   
Willowdale, Newtonbrook                                        2   
Woburn                                                         4   
Woodbine Heights                                               5   
York Mills West                                                2   

                                                 Venue Category  
Neighborhood                                                     
Agincourt                                                     4  
Alderwood, Long Branch                                        7 

In [19]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 256 uniques categories.


## Analysis of Neighbourhoods 

In [20]:
# one hot encoding
TO_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
TO_onehot['Neighborhood'] = Toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [TO_onehot.columns[-1]] + list(TO_onehot.columns[:-1])
TO_onehot = TO_onehot[fixed_columns]

TO_onehot.head()

Yoga Studio  Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0            0                  0               0                  0        0   
1            0                  0               0                  0        0   
2            0                  0               0                  0        0   
3            0                  0               0                  0        0   
4            0                  0               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  ...  Truck Stop  Vegetarian / Vegan Restaurant  \
0                 0  ...           0                              0   
1                 0  ...           0                              0   
2                 0  ...           0                              0   
3                 0  ...           0                              0   
4                 0  ...           0                              0   

   Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                 0            0                      0                0   
1                 0            0                      0                0   
2                 0            0                      0                0   
3                 0            0                      0                0   
4                 0            0                      0                0   

   Wine Bar  Wine Shop  Wings Joint  Women's Store  
0         0          0            0              0  
1         0          0            0              0  
2         0          0            0              0  
3         0          0            0              0  
4         0          0            0              0  

[5 rows x 256 columns]

Now we can group the neighbourhoods and take the mean frequency of each venue

In [21]:
TO_grouped = TO_onehot.groupby('Neighborhood').mean().reset_index()
TO_grouped

Neighborhood  Yoga Studio  \
0                                         Agincourt          0.0   
1                            Alderwood, Long Branch          0.0   
2   Bathurst Manor, Wilson Heights, Downsview North          0.0   
3                                   Bayview Village          0.0   
4                 Bedford Park, Lawrence Manor East          0.0   
..                                              ...          ...   
94                                  Willowdale West          0.0   
95                          Willowdale, Newtonbrook          0.0   
96                                           Woburn          0.0   
97                                 Woodbine Heights          0.0   
98                                  York Mills West          0.0   

    Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0                 0.0             0.0                0.0      0.0   
1                 0.0             0.0                0.0      0.0   
2                 0.0             0.0                0.0      0.0   
3                 0.0             0.0                0.0      0.0   
4                 0.0             0.0                0.0      0.0   
..                ...             ...                ...      ...   
94                0.0             0.0                0.0      0.0   
95                0.0             0.0                0.0      0.0   
96                0.0             0.0                0.0      0.0   
97                0.0             0.0                0.0      0.0   
98                0.0             0.0                0.0      0.0   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  ...  \
0                  0.0           0.0             0.0              0.0  ...   
1                  0.0           0.0             0.0              0.0  ...   
2                  0.0           0.0             0.0              0.0  ...   
3                  0.0           0.0             0.0              0.0  ...   
4                  0.0           0.0             0.0              0.0  ...   
..                 ...           ...             ...              ...  ...   
94                 0.0           0.0             0.0              0.0  ...   
95                 0.0           0.0             0.0              0.0  ...   
96                 0.0           0.0             0.0              0.0  ...   
97                 0.0           0.0             0.0              0.0  ...   
98                 0.0           0.0             0.0              0.0  ...   

    Truck Stop  Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0          0.0                            0.0               0.0          0.0   
1          0.0                            0.0               0.0          0.0   
2          0.0                            0.0               0.0          0.0   
3          0.0                            0.0               0.0          0.0   
4          0.0                            0.0               0.0          0.0   
..         ...                            ...               ...          ...   
94         0.0                            0.0               0.0          0.0   
95         0.0                            0.0               0.0          0.0   
96         0.0                            0.0               0.0          0.0   
97         0.0                            0.0               0.0          0.0   
98         0.0                            0.0               0.0          0.0   

    Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  Wings Joint  \
0                     0.0              0.0       0.0        0.0          0.0   
1                     0.0              0.0       0.0        0.0          0.0   
2                     0.0              0.0       0.0        0.0          0.0   
3                     0.0              0.0       0.0        0.0          0.0   
4                     0.0              0.0       0.0        0.0          0.0   
..             

In [22]:
TO_grouped.shape

(99, 256)

#### Let's print each neighborhood along with the top 5 most common venues

In [23]:
num_top_venues = 5

for hood in TO_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = TO_grouped[TO_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0                     Lounge  0.25
1               Skating Rink  0.25
2  Latin American Restaurant  0.25
3             Breakfast Spot  0.25
4                Yoga Studio  0.00


----Alderwood, Long Branch----
            venue  freq
0     Pizza Place  0.29
1        Pharmacy  0.14
2    Skating Rink  0.14
3             Pub  0.14
4  Sandwich Place  0.14


----Bathurst Manor, Wilson Heights, Downsview North----
                       venue  freq
0                       Bank  0.10
1                Coffee Shop  0.10
2                Pizza Place  0.05
3                  Gift Shop  0.05
4  Middle Eastern Restaurant  0.05


----Bayview Village----
                 venue  freq
0                 Café  0.25
1  Japanese Restaurant  0.25
2                 Bank  0.25
3   Chinese Restaurant  0.25
4               Museum  0.00


----Bedford Park, Lawrence Manor East----
                     venue  freq
0              Coffee Shop  0.10
1               R

            venue  freq
0     Coffee Shop  0.17
1            Café  0.04
2           Hotel  0.04
3  Scenic Lookout  0.03
4     Pizza Place  0.03


----High Park, The Junction South----
                       venue  freq
0                        Bar  0.11
1                       Café  0.11
2         Mexican Restaurant  0.11
3       Fast Food Restaurant  0.06
4  Cajun / Creole Restaurant  0.06


----Hillcrest Village----
                      venue  freq
0               Golf Course   0.2
1      Fast Food Restaurant   0.2
2  Mediterranean Restaurant   0.2
3                      Pool   0.2
4                   Dog Run   0.2


----Humber Summit----
                 venue  freq
0          Pizza Place   0.5
1                  Gym   0.5
2                Motel   0.0
3  Martial Arts School   0.0
4       Medical Center   0.0


----Humberlea, Emery----
                 venue  freq
0       Baseball Field   1.0
1          Yoga Studio   0.0
2     Malay Restaurant   0.0
3  Martial Arts School   0.0
4   

4        Pizza Place  0.05


----Toronto Dominion Centre, Design Exchange----
              venue  freq
0       Coffee Shop  0.13
1              Café  0.08
2             Hotel  0.05
3    Sandwich Place  0.05
4  Asian Restaurant  0.04


----University of Toronto, Harbord----
            venue  freq
0            Café  0.14
1     Coffee Shop  0.08
2  Sandwich Place  0.08
3             Bar  0.06
4             Pub  0.06


----Victoria Village----
                   venue  freq
0            Coffee Shop  0.25
1           Intersection  0.25
2           Hockey Arena  0.25
3  Portuguese Restaurant  0.25
4            Yoga Studio  0.00


----West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale----
                 venue  freq
0           Print Shop   1.0
1          Yoga Studio   0.0
2                Motel   0.0
3  Martial Arts School   0.0
4       Medical Center   0.0


----Westmount----
                       venue  freq
0                Pizza Place  0.22
1         Chinese Resta

Now we can store this data in a pandas dataframe

In [24]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [25]:
import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = TO_grouped['Neighborhood']

for ind in np.arange(TO_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(TO_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East           Coffee Shop   

  2nd Most Common Venue      3rd Most Common Venue    4th Most Common Venue  \
0          Skating Rink  Latin American Restaurant           Breakfast Spot   
1              Pharmacy               Skating Rink                      Pub   
2           Coffee Shop                Pizza Place                Gift Shop   
3   Japanese Restaurant                       Bank       Chinese Restaurant   
4            Restaurant             Sandwich Place  Comfort Food Restaurant   

       5th Most Common Venue 6th Most Common Venue     7th Most Common Venue  \
0                Yoga Studio   Moroccan Restaurant  Mediterranean Restaurant   
1             Sandwich Place           Coffee Shop               Men's Store   
2  Middle Eastern Restaurant           Bridal Shop                  Pharmacy   
3                     Museum                 Motel       Moroccan Restaurant   
4                       Café               Butcher              Liquor Store   

  8th Most Common Venue            9th Most Common Venue  \
0           Men's Store                    Metro Station   
1         Metro Station               Mexican Restaurant   
2         Shopping Mall                   Sandwich Place   
3   Monument / Landmark  Molecular Gastronomy Restaurant   
4       Thai Restaurant                              Pub   

       10th Most Common Venue  
0          Mexican Restaurant  
1         Moroccan Restaurant  
2            Sushi Restaurant  
3  Modern European Restaurant  
4          Italian Restaurant

## Finally, lets use K-means clustering to cluster the neighborhoods

In [26]:
# set number of clusters
kclusters = 5

TO_grouped_clustering = TO_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(TO_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 2, 2, 2, 2, 2, 2, 2, 2, 0])

In [27]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

TO_merged = joindf

# merge grouped data with manhattan_data to add latitude/longitude for each neighborhood
TO_merged = TO_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

In [31]:
TO_merged['Cluster Labels'] = TO_merged['Cluster Labels'].fillna(0).astype(int)

TO_merged.head() # check the last columns!

index           Borough                      Neighborhood   Latitude  \
0   M3A        North York                         Parkwoods  43.753259   
1   M4A        North York                  Victoria Village  43.725882   
2   M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3   M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4   M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  Cluster Labels 1st Most Common Venue   2nd Most Common Venue  \
0 -79.329656               0                 Hotel    Fast Food Restaurant   
1 -79.315572               2           Coffee Shop            Intersection   
2 -79.360636               2           Coffee Shop                  Bakery   
3 -79.464763               2        Clothing Store  Furniture / Home Store   
4 -79.389494               2           Coffee Shop        Sushi Restaurant   

  3rd Most Common Venue  4th Most Common Venue      5th Most Common Venue  \
0                  Park      Food & Drink Shop  Middle Eastern Restaurant   
1          Hockey Arena  Portuguese Restaurant                Yoga Studio   
2                   Pub                   Park                       Café   
3              Boutique            Coffee Shop      Vietnamese Restaurant   
4         Burrito Place                    Bar                       Bank   

             6th Most Common Venue       7th Most Common Venue  \
0  Molecular Gastronomy Restaurant  Modern European Restaurant   
1                Mobile Phone Shop         Moroccan Restaurant   
2                   Sandwich Place                        Bank   
3                     Carpet Store          Miscellaneous Shop   
4                             Café          Mexican Restaurant   

  8th Most Common Venue            9th Most Common Venue  \
0     Mobile Phone Shop               Miscellaneous Shop   
1   Monument / Landmark  Molecular Gastronomy Restaurant   
2            Restaurant                Electronics Store   
3           Yoga Studio       Modern European Restaurant   
4                 Diner                          Theater   

       10th Most Common Venue  
0                 Yoga Studio  
1  Modern European Restaurant  
2              Chocolate Shop  
3         Moroccan Restaurant  
4              Sandwich Place

Now we can map this data

In [32]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(TO_merged['Latitude'], TO_merged['Longitude'], TO_merged['Neighborhood'], TO_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examining Clusters

Cluster 1

In [34]:
TO_merged.loc[TO_merged['Cluster Labels'] == 0, TO_merged.columns[[1] + list(range(5, TO_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0               North York               0                 Hotel   
5                Etobicoke               0                   NaN   
10              North York               0    Italian Restaurant   
21                    York               0                  Park   
35  East York/East Toronto               0                  Park   
40              North York               0               Airport   
45              North York               0                   NaN   
46              North York               0         Grocery Store   
49              North York               0      Basketball Court   
52              North York               0                  Park   
61         Central Toronto               0              Bus Line   
64                    York               0                   NaN   
66              North York               0                  Park   
68         Central Toronto               0         Jewelry Store   
77               Etobicoke               0        Sandwich Place   
85             Scarborough               0            Playground   
91        Downtown Toronto               0                  Park   
95             Scarborough               0                   NaN   
98               Etobicoke               0                  Pool   

   2nd Most Common Venue 3rd Most Common Venue       4th Most Common Venue  \
0   Fast Food Restaurant                  Park           Food & Drink Shop   
5                    NaN                   NaN                         NaN   
10                  Park            Playground                   Pet Store   
21         Women's Store                  Pool   Middle Eastern Restaurant   
35          Intersection     Convenience Store                 Yoga Studio   
40                  Park           Yoga Studio   Middle Eastern Restaurant   
45                   NaN                   NaN                         NaN   
46                  Bank                  Park               Shopping Mall   
49                Bakery                  Park  Construction & Landscaping   
52           Yoga Studio                 Motel         Martial Arts School   
61           Swim School                  Park            Business Service   
64                   NaN                   NaN                         NaN   
66     Convenience Store           Yoga Studio   Middle Eastern Restaurant   
68                  Park                 Trail            Sushi Restaurant   
77                  Park     Mobile Phone Shop                 Yoga Studio   
85                  Park      Asian Restaurant                Intersection   
91            Playground                 Trail   Middle Eastern Restaurant   
95                   NaN                   NaN                         NaN   
98                 River                  Park                 Yoga Studio   

        5th Most Common Venue            6th Most Common Venue  \
0   Middle Eastern Restaurant  Molecular Gastronomy Restaurant   
5                         NaN                              NaN   
10        Moroccan Restaurant              Martial Arts School   
21        Monument / Landmark  Molecular Gastronomy Restaurant   
35         Miscellaneous Shop              Monument / Landmark   
40        Monument / Landmark  Molecular Gastronomy Restaurant   
45                        NaN                              NaN   
46                Yoga Studio               Miscellaneous Shop   
49  Middle Eastern Restaurant  Molecular Gastronomy Restaurant   
52             Medical Center         Mediterranean Restaurant   
61                Yoga Studio                Mobile Phone Shop   
64                        NaN                              NaN   
66        Monument / Landmark  Molecular Gastronomy Restaurant   
68                Yoga Studio        Middle Eastern Restaurant   
77        Martial Arts School                   Medical Center   
85         Persian Restaurant            

Cluster 2

In [36]:
TO_merged.loc[TO_merged['Cluster Labels'] == 1, TO_merged.columns[[1] + list(range(5, TO_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
32  Scarborough               1            Playground                 Motel   

   3rd Most Common Venue 4th Most Common Venue     5th Most Common Venue  \
32   Martial Arts School        Medical Center  Mediterranean Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
32           Men's Store         Metro Station    Mexican Restaurant   

        9th Most Common Venue 10th Most Common Venue  
32  Middle Eastern Restaurant     Miscellaneous Shop

Cluster 3

In [37]:
TO_merged.loc[TO_merged['Cluster Labels'] == 2, TO_merged.columns[[1] + list(range(5, TO_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1               North York               2           Coffee Shop   
2         Downtown Toronto               2           Coffee Shop   
3               North York               2        Clothing Store   
4             Queen's Park               2           Coffee Shop   
7               North York               2   Japanese Restaurant   
..                     ...             ...                   ...   
96        Downtown Toronto               2           Coffee Shop   
97        Downtown Toronto               2           Coffee Shop   
99        Downtown Toronto               2   Japanese Restaurant   
100  East Toronto Business               2    Light Rail Station   
102              Etobicoke               2                Bakery   

      2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
1              Intersection          Hockey Arena  Portuguese Restaurant   
2                    Bakery                   Pub                   Park   
3    Furniture / Home Store              Boutique            Coffee Shop   
4          Sushi Restaurant         Burrito Place                    Bar   
7                       Gym          Dessert Shop   Caribbean Restaurant   
..                      ...                   ...                    ...   
96                   Bakery                  Café             Restaurant   
97           Sandwich Place                  Café                  Hotel   
99         Sushi Restaurant           Coffee Shop             Restaurant   
100             Pizza Place               Brewery          Garden Center   
102  Thrift / Vintage Store           Wings Joint         Discount Store   

     5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
1              Yoga Studio     Mobile Phone Shop        Moroccan Restaurant   
2                     Café        Sandwich Place                       Bank   
3    Vietnamese Restaurant          Carpet Store         Miscellaneous Shop   
4                     Bank                  Café         Mexican Restaurant   
7                     Café    Mexican Restaurant  Middle Eastern Restaurant   
..                     ...                   ...                        ...   
96                     Pub    Italian Restaurant                Pizza Place   
97                    Bank                   Gym        Japanese Restaurant   
99                 Gay Bar     Indian Restaurant                        Pub   
100          Auto Workshop            Restaurant              Burrito Place   
102          Tanning Salon            Kids Store                        Gym   

    8th Most Common Venue            9th Most Common Venue  \
1     Monument / Landmark  Molecular Gastronomy Restaurant   
2              Restaurant                Electronics Store   
3             Yoga Studio       Modern European Restaurant   
4                   Diner                          Theater   
7      Miscellaneous Shop                Mobile Phone Shop   
..                    ...                              ...   
96         Sandwich Place             Caribbean Restaurant   
97          Deli / Bodega                 Sushi Restaurant   
99          Burrito Place                              Gym   
100            Skate Park                           Garden   
102         Grocery Store                Convenience Store   

         10th Most Common Venue  
1    Modern European Restaurant  
2                Chocolate Shop  
3           Moroccan Restaurant  
4                Sandwich Place  
7                        Museum  
..                          ...  
96                       Market  
97                   Restaurant  
99         Fast Food Restaurant  
100                        Park  
102             Supplement Shop  

[80 rows x 12 columns]

Cluster 4

In [38]:
TO_merged.loc[TO_merged['Cluster Labels'] == 3, TO_merged.columns[[1] + list(range(5, TO_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
57   North York               3        Baseball Field           Yoga Studio   
101   Etobicoke               3        Baseball Field           Yoga Studio   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
57       Malay Restaurant   Martial Arts School        Medical Center   
101      Malay Restaurant   Martial Arts School        Medical Center   

        6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
57   Mediterranean Restaurant           Men's Store         Metro Station   
101  Mediterranean Restaurant           Men's Store         Metro Station   

    9th Most Common Venue     10th Most Common Venue  
57     Mexican Restaurant  Middle Eastern Restaurant  
101    Mexican Restaurant  Middle Eastern Restaurant

Cluster 5

In [39]:
TO_merged.loc[TO_merged['Cluster Labels'] == 4, TO_merged.columns[[1] + list(range(5, TO_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough               4  Fast Food Restaurant           Yoga Studio   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
6      Malay Restaurant   Martial Arts School        Medical Center   

      6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6  Mediterranean Restaurant           Men's Store         Metro Station   

  9th Most Common Venue     10th Most Common Venue  
6    Mexican Restaurant  Middle Eastern Restaurant